docs/architecture.md

```diff
@@ -15,139 +15,207 @@ Prometheus instances scrape samples from various targets and then push them to C
 
-Distributors use consistent hashing, in conjunction with the (configurable) replication factor, to determine *which* instances of the ingester service receive each sample.
+The HA Tracker requires a key-value (KV) store to coordinate which is the currently elected primary replica from which the distributor accepts the samples. Samples with one or neither of the replica and cluster labels are accepted by default and never deduplicated.
 
-The hash itself is based on one of two schemes:
+The supported KV stores for the HA tracker are:
 
-1. The metric name and tenant ID
-2. All the series labels and tenant ID
+* [Consul](https://www.consul.io)
+* [Etcd](https://etcd.io)
+
+For more information, please check out the [Config for sending HA pairs data to Cortex](guides/ha-pair-handling.md) documentation.
+
+#### Hashing
 
-The trade-off associated with the latter is that writes are more balanced but they must involve every ingester in each query.
+Distributors use consistent hashing, in conjunction with a configurable replication factor, to determine which instances of the ingester service should receive a given series.
 
-> This hashing scheme was originally chosen to reduce the number of required ingesters on the query path. The trade-off, however, is that the write load on the ingesters is less even.
+Cortex supports two hashing strategies:
+
+1. Hash the metric name and tenant ID (default)
+2. Hash the metric name, labels and tenant ID (recommended, enabled with `-distributor.shard-by-all-labels=true`)
+
+The trade-off associated with the latter is that writes are more balanced across ingesters but each query needs to talk to any ingester since a metric could be spread across any ingester given different label sets. The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying all ingesters.
 
 #### The hash ring
 
-A consistent hash ring is stored in [Consul](https://www.consul.io/) as a single key-value pair, with the ring data structure also encoded as a [Protobuf](https://developers.google.com/protocol-buffers/) message. The consistent hash ring consists of a list of tokens and ingesters. Hashed values are looked up in the ring; the replication set is built for the closest unique ingesters by token. One of the benefits of this system is that adding and remove ingesters results in only 1/_N_ of the series being moved (where _N_ is the number of ingesters).
```

Comment:
In my opinion, this section is one slightly more complex pieces of Cortex. You've done a fantastic job trying to expand on it, but I'm not sure about the way it reads.

If you feel strongly about it we can leave it as is, if you have doubts - I'm happy to take a stab and try to reword it and perhaps add a small diagram to it. What do you think?

Reply:
I never feel very strong about something, and I'm always open to discussion 😉 That being said, I would appreciate if could express the same concept in a easier way.

About the diagram, sure. I also thought about it, but I didn't (yet) because I'm not sure how to draw a diagram in a way which will be easily editable by other contributors in the future (ie. which tool to use? where should we share the sources?).

Reply:
I tend to use [mermaid](https://mermaidjs.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbkFbQ2hyaXN0bWFzXSAtLT58R2V0IG1vbmV5fCBCKEdvIHNob3BwaW5nKVxuQiAtLT4gQ3tMZXQgbWUgdGhpbmt9XG5DIC0tPnxPbmV8IERbTGFwdG9wXVxuQyAtLT58VHdvfCBFW2lQaG9uZV1cbkMgLS0-fFRocmVlfCBGW2ZhOmZhLWNhciBDYXJdXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9fQ) for that exact same reasons. The source code is extremely simple which makes it easy to check it into version control and they have a live editor which means anyone can copy/paste then start editing right away.

Let's merge this as is and I'll create a follow-up issue assigned to me to come up with something.

Reply:
I have created #1949 to record this conversation. Let me know what you think.

Reply:
I think it's good not blocking this PR with the improved ring description. I will comment the issue.